lib/ts/recipe/session/sessionAuth.tsx

```diff
@@ -98,79 +111,75 @@ const SessionAuth: React.FC<PropsWithChildren<Props>> = ({ children, ...props })
 
     const setInitialContextAndMaybeRedirect = useCallback(
         async (toSetContext: SessionContextType) => {
-            // if this component is unmounting, or the context has already
-            // been set, then we don't need to proceed...
+            if (toSetContext.loading === true) {
+                // We should not be updating the context to loading
+                throw new Error("Should never come here");
+            }
+
             if (!toSetContext.doesSessionExist && props.requireAuth === true) {
-                props.redirectToLogin();
+                redirectToLogin();
             } else {
-                if (context === undefined) {
-                    setContext(toSetContext);
-                }
+                setContext((os) => (os.loading === false ? os : toSetContext));
             }
         },
-        [props]
+        [props.requireAuth, redirectToLogin]
     );
 
     useOnMountAPICall(buildContext, setInitialContextAndMaybeRedirect);
 
     // subscribe to events on mount
     useEffect(() => {
-        function onHandleEvent(event: RecipeEventWithSessionContext) {
+        async function onHandleEvent(event: RecipeEventWithSessionContext) {
             switch (event.action) {
+                // We intentionally fall through as they are all handled the same way.
                 case "SESSION_CREATED":
-                    setContext(event.sessionContext);
-                    return;
                 case "REFRESH_SESSION":
-                    setContext(event.sessionContext);
-                    return;
                 case "ACCESS_TOKEN_PAYLOAD_UPDATED":
-                    setContext(event.sessionContext);
-                    return;
                 case "SIGN_OUT":
```

Comment:
do we need to do claim validation if this is the true as well?

Reply:
(Maybe make this similar to UNAUTHORISED case, but without calling the callbacks